File: py/src/domain/value_objects/money.py
from src.domain.value_objects.currency import Currency

class Money:
    def __init__(self, amount: float | int, currency: Currency):
        # Allow ints or floats, but validate decimals
        amount_val = float(amount)
        if amount_val < 0:
            raise ValueError("Money amount cannot be negative.")
        
        # Validate maximum of 2 decimal places to avoid floating point precision issues
        if round(amount_val, 2) != amount_val:
            raise ValueError("Money amount cannot have more than 2 decimal places.")
            
        self._amount = amount_val
        self._currency = currency

    @property
    def amount(self) -> float:
        return self._amount

    @property
    def currency(self) -> Currency:
        return self._currency

    @staticmethod
    def zero(currency: Currency) -> 'Money':
        return Money(0, currency)

    def add(self, other: 'Money') -> 'Money':
        self._check_same_currency(other)
        # Prevent floating point anomalies by doing integer round-math
        new_amount = round(self._amount + other.amount, 2)
        return Money(new_amount, self._currency)

    def subtract(self, other: 'Money') -> 'Money':
        self._check_same_currency(other)
        if self._amount < other.amount:
            raise ValueError("Insufficient funds for subtraction.")
        new_amount = round(self._amount - other.amount, 2)
        return Money(new_amount, self._currency)

    def is_greater_than_or_equal(self, other: 'Money') -> bool:
        self._check_same_currency(other)
        return self._amount >= other.amount

    def equals(self, other: 'Money') -> bool:
        if not isinstance(other, Money):
            return False
        return self._amount == other.amount and self._currency.equals(other.currency)

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Money):
            return NotImplemented
        return self.equals(other)

    def _check_same_currency(self, other: 'Money') -> None:
        if not self._currency.equals(other.currency):
            raise ValueError(
                f"Currency mismatch: cannot perform operation between {self._currency.code} and {other.currency.code}"
            )

    def __repr__(self) -> str:
        return f"Money({self._amount}, {self._currency})"

File: js/src/domain/value-objects/Money.ts
import { Currency } from "./Currency";

export class Money {
  constructor(
    public readonly amount: number,
    public readonly currency: Currency
  ) {
    if (amount < 0) {
      throw new Error("Money amount cannot be negative.");
    }
    // Prevent floating point issue by validating up to 2 decimal places
    if (Number.isFinite(amount) && Math.round(amount * 100) / 100 !== amount) {
      throw new Error("Money amount cannot have more than 2 decimal places.");
    }
  }

  public static zero(currency: Currency): Money {
    return new Money(0, currency);
  }

  public add(other: Money): Money {
    this.checkSameCurrency(other);
    // Use integer math to prevent floating point issues
    const newAmount = (Math.round(this.amount * 100) + Math.round(other.amount * 100)) / 100;
    return new Money(newAmount, this.currency);
  }

  public subtract(other: Money): Money {
    this.checkSameCurrency(other);
    if (this.amount < other.amount) {
      throw new Error("Insufficient funds for subtraction.");
    }
    const newAmount = (Math.round(this.amount * 100) - Math.round(other.amount * 100)) / 100;
    return new Money(newAmount, this.currency);
  }

  public isGreaterThanOrEqual(other: Money): boolean {
    this.checkSameCurrency(other);
    return this.amount >= other.amount;
  }

  public equals(other: Money): boolean {
    return this.amount === other.amount && this.currency.equals(other.currency);
  }

  private checkSameCurrency(other: Money): void {
    if (!this.currency.equals(other.currency)) {
      throw new Error(
        `Currency mismatch: cannot perform operation between ${this.currency.code} and ${other.currency.code}`
      );
    }
  }
}
